flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/codegen/CodeGenUtils.scala

```diff
@@ -556,7 +556,7 @@ object CodeGenUtils {
         s"$rowTerm.setNullAt($indexTerm)"
       }
 
-      if (fieldType.isNullable) {
+      if (fieldType.isNullable || fieldExpr.nullTerm.startsWith("castRuleResultIsNull")) {
```

Comment:
IMHO it is not good to rely on var namings
rather it's better to check properties
why can't we pass a corresponding arg to `org.apache.flink.table.planner.codegen.calls.ScalarOperatorGens$#generateCast` for `nullOnFailure` ?